src/coreclr/_build-commons.sh

```diff
@@ -414,12 +414,12 @@ if [ "$__VerboseBuild" = 1 ]; then
 fi
 
 # Set default clang version
-if [ "$__ClangMajorVersion" = 0 ] && [ "$__ClangMinorVersion" = 0 ]; then
+if [ "$__CompilerMajorVersion" = 0 ] && [ "$__CompilerMinorVersion" = 0 ]; then
     if [ "$__BuildArch" = "arm" ] || [ "$__BuildArch" = "armel" ]; then
-        __ClangMajorVersion=5
-        __ClangMinorVersion=0
+        __CompilerMajorVersion=5
```

Comment:
Sure. Currently, validation, fallback and selection logic for compiler is scattered in multiple places, and it is hard to wrap our head around the whole picture. I have tried to consolidate all pieces into `eng/common/cross/gen-buildsys.sh`, in the second commit. It goes like this:

* Compiler introspection effective:
  * initially no version is set (changed 0.0 to null)
  * if user has not specified the version:
    * `gen-buildsys.sh` will check if any of the supported versions exist in PATH:
      * clang: 9, 8, 7, 6.0, 5.0, 4.0, 3.9, 3.8, 3.7, 3.6, 3.5
      * gcc: 9, 8, 7, 6, 5, 4.9
    <sub># note: gcc (all versions) and clang versions higher than 6 do not have minor version in file name, if it is zero.</sub>
    * if a supported version exists, it validates "clang for arm/armel version < 5" case; where it further checks if there is any version-less clang in PATH, falls back to it with warning. Otherwise, error out that there is no usable version of clang found.
    * if there is no supported version found, check for any version-less occurrence of desired compiler is in PATH, falls back to it with warning (except in case of Darwin warning is skipped; because AppleClang always has version-less filename). Otherwise, error out that there is no usable version of desired compiler found.
  * if user had specified the version (e.g. `-clang`, `-clang9`, `-clang3.6`, `-gcc7`):
    * check if that version really exists. if not, check for any version-less occurrence of desired compiler in PATH, fallback to it with warning. Otherwise, error out that there is no usable version of desired compiler.
* Compiler introspection will be skipped when `CLR_CC` environment variable is set: if user wishes to build coreclr and libraries using custom/portable clang/gcc toolchain, it is possible. The environment variables that build honors are: `CLR_CC`, `CLR_CXX`, `CLR_AR`, `CLR_LINK`, `CLR_NM`, `CLR_OBJDUMP`, `CLR_OBJCOPY` and `CLR_RANLIB`.
  * in this case, only validate and error out if the file which `$CLR_CC` points to doe not exist.